### PR TITLE
Fix legacy env var support with cross

### LINKED ISSUE #3969: Meson's tool env var interpretation clashes with Autoconf's, leading to disto packaging issue
With Autoconf (along with ideomatic macros):
 - `<TOOL>_FOR_BUILD` affects the build platform
 - `<TOOL>` affects the host platform
 - `<TOOL>_FOR_TARGET` affects the target platform.

For example, if I am building a Canadian cross GCC:
 - `CC_FOR_BUILD` builds tools on the fly used to generate some repetitive C code. 
 - `CC` builds the cross compiler itself
 - `CC_FOR_TARGET` builds `libgcc` and other runtime libraries, since the newly-built compiler cannot be run to build them.

Meson however assumes `CC`, `AS` and friends always target the native platform, whatever that is. Notice how in terms of "build", "host" and "target", rather than "native", this means the meaning of these environment variables in Meson is conditional on whether we are cross compiling or not. `CC` is explicitly used in the native case for building things morally for the host platform (i.e. stuff that will be installed), but not in the cross case. 

In the distro I package use, NixOS, We define the variables according to Autoconf https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/bintools-wrapper/setup-hook.sh#L66 here. (Granted there's a number of abstractions going on so I don't expect that to be clear.) This works in the fast majority of cases, but creates issues for cross-compiling our packages that use Meson, even though we *also* create a cross file just for meson. It sees our `CC`, targeting the host platform, and tries to use that rather than `CC_FOR_BUILD` for building code gen tools. This fails as the resulting binaries cannot be run at build time.

At a minimum, this issue would be solved if Meson would ignore `CC` for building code gen tools in the cross case. It would be nicer still if Meson would read `<TOOL>_FOR_BUILD` rather than `<TOOL>` in that cross case. Even better would be using `<TOOL>_FOR_BUILD` for building code gen stuff in *all* cases: this would make the native in cross cases more consistent. (It would also help with e.g. optimizing release builds but not codegen tools used for release buidls, FWIW.)

Since environment variables are not an idiomatic way to give Meson information, it would also be nice to specify the tools with (something like) the cross file. I opened https://github.com/mesonbuild/meson/issues/3282 for this purpose, but it got no response. Note that this is very analogous to specifying `build_machine` information more broadly, I proposed in https://github.com/mesonbuild/meson/pull/3689 but seems to be approaching rejected.

CC @Dezgeg @jtojnar 

-------

So since I opened this issue, we've had two big changes internally:

1. #4216 native file support lands
2. #4445 all binary input (environment variables and config files) goes through `BinaryTable` class.

This makes solving this a lot easier. I suggest:

1. ~~Native files have a `[binaries]` section, which functions in the obvious way (like the env vars today)~~ done, my mistake.
2. `FOO_FOR_BUILD` always overrides `FOO` for the binaries for build (`environment.binaries.build`).
3. `FOO_FOR_BUILD` is the only way to specify for-build binaries for cross builds. Plain `FOO` only controls host binaries for cross builds.

In particular, 1 and 2 and land in isolation, and are *not* breaking changes. Only 3 is, and by then the idiomatic way to control binaries would be the native file anyways.

`CFLAGS` and friends are more complicated, as the code paths are still wildly different.

### PR DESCRIPTION
**Pull request by Ericson2314**
This fixes us to match Autoconf's behavior; see [the issue](https://github.com/mesonbuild/meson/issues/3969). It does however still keep it so the cross is the only way to control the host platform in cross builds.

CC @dcbaker

Fixes #3969

### CODE REVIEW

File: mesonbuild/modules/python3.py [hunk 1]
```diff
@@ -48,7 +48,10 @@ def extension_module(self, interpreter, state, args, kwargs):
 
     @noKwargs
     def find_python(self, state, args, kwargs):
-        command = state.environment.binaries.host.lookup_entry('python3')
+        command = state.environment.binaries.host.lookup_entry(
+            MachineChoice.HOST,
```
Comment (ignatenkobrain) [inline mesonbuild/modules/python3.py]: __[Flake8]__ undefined name 'MachineChoice' ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6363#issue-4963464))

<details>

| Rule |
| - |
| F821 |

</details>
<sub>posted by Sider</sub>

File: mesonbuild/linkers.py [hunk 2]
```diff
@@ -233,7 +234,24 @@ def evaluate_rpath(p: str, build_dir: str, from_dir: str) -> str:
...
+    """Mixin reading LDFLAGS from the environment."""
+
+    @staticmethod
+    def get_args_from_envvars(for_machine: mesonlib.MachineChoice,
+                              is_cross: bool) -> typing.List[str]:
+        env_var = 'LDFLAGS'
+        env_var, raw_value = get_env_var_with_key(for_machine, is_cross, env_var)
+        if raw_value is not None:
+            flags = mesonlib.split_args(raw_value)
+            log_var(env_var, flags)
```
Comment (ignatenkobrain) [inline mesonbuild/linkers.py]: __[Flake8]__ undefined name 'log_var' ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6363#issue-4963463))

<details>

| Rule |
| - |
| F821 |

</details>
<sub>posted by Sider</sub>

File: mesonbuild/compilers/mixins/islinker.py [hunk 3]
```diff
@@ -24,23 +24,13 @@
 import typing
 
 from ... import mesonlib
+from ...linkers import DynamicLinker, LinkerEnvVarsMixin # for re-export
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/mixins/islinker.py]: __[Flake8]__ '...linkers.DynamicLinker' imported but unused ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6363#issue-4963462))

<details>

| Rule |
| - |
| F401 |

</details>
<sub>posted by Sider</sub>

**Comment by jpakkane** [comment]
Cross compilation is never, ever, *ever* going to look into environment variables. This is a design point since day one.

The aim in Meson has always been to use fewer environment variables, not more of them.

**Comment by Ericson2314** [comment]
@jpakkane At first as I was erked, but then I realized that I titled this PR rather poorly. The big point of #3969 is not to use more environment variables, but using them currently. We currently *conflict* Autoconf's behavior, which is a huge pain for distros such as NixOS and Exherbo that care about cross as it means we/they have to take a bunch of steps to suppress env vars.

File: mesonbuild/compilers/compilers.py [hunk 4]
```diff
@@ -1128,37 +1135,28 @@ def get_largefile_args(compiler):
...
     env_compile_flags = os.environ.get(cflags_mapping[lang])
-    log_var(cflags_mapping[lang], env_compile_flags)
+    log_env_var(cflags_mapping[lang], env_compile_flags)
     if env_compile_flags is not None:
         compile_flags += split_args(env_compile_flags)
 
     # Link flags (same for all languages)
     if lang in languages_using_ldflags:
-        # This is duplicated between the linkers, but I'm not sure how else
-        # to handle this
```
Comment (Ericson2314) [inline mesonbuild/compilers/compilers.py]: @dcbaker Now fixed!

File: mesonbuild/dependencies/base.py [hunk 5]
```diff
@@ -1110,8 +1114,12 @@ def __init__(self, name: str, environment: Environment, kwargs, language: str =
             cm_args.append('-DCMAKE_MODULE_PATH=' + ';'.join(cm_path))
 
         pref_path = self.env.coredata.builtins_per_machine[self.for_machine]['cmake_prefix_path'].value
-        if 'CMAKE_PREFIX_PATH' in os.environ:
-            env_pref_path = os.environ['CMAKE_PREFIX_PATH'].split(os.pathsep)
+        env_pref_path = get_env_var(
+            self.for_machine,
+            self.env.is_cross_build(),
+            'CMAKE_PREFIX_PATH')
+        if cmake_prefix_path is not None:
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__ undefined name 'cmake_prefix_path' ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6363#issue-4973968))

<details>

| Rule |
| - |
| F821 |

</details>
<sub>posted by Sider</sub>

File: run_unittests.py [hunk 6]
```diff
@@ -5837,13 +5840,34 @@ def test_noncross_options(self):
...
         crossfile.flush()
         self.meson_cross_file = crossfile.name
+
+        # TODO should someday be explicit about build platform only here
+        self.init(testdir)
+
+    def test_identity_cross_env(self):
+        testdir = os.path.join(self.unit_test_dir, '61 identity cross')
+        env = {
+          'CC_FOR_BUILD': '"' + os.path.join(testdir, 'build_wrapper.py') + '"',
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__ continuation line under-indented for hanging indent ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6363#issue-4973969))

<details>

| Rule |
| - |
| E121 |

</details>
<sub>posted by Sider</sub>

**Comment by Ericson2314** [comment]
Furthermore, I've made Meson consistently log it's use of env vars so the user has a better idea what's going on. and they know what information that need to instead convey in their config file.

**Comment by dcbaker** [comment]
I'm not really sure how I feel about this. I agree with @jpakkane that using environment variables is rather annoying, and I'd rather not do them. It's also changing the meaning (which is going to make 50% of people happy and make 50% annoyed). Honestly I'd prefer to just drop env var support if we're going to change anything, but that ship sailed already...

**Comment by Ericson2314** [comment]
@dcbaker 

> It's also changing the meaning (which is going to make 50% of people happy and make 50% annoyed).

It's important to note that for native builds everything will work the same.

For users doing cross builds, yes, there would be some breakage. But again people should really take the opportunity to switch to using native files.

The big beneficiaries are distros that support cross, and I'll be quite happy to delete the crap we have today in NixOS and am happy to pitch in with Exherbo too.

Maybe we could agree to first land everything we need to generate config files based on flags and env vars, so the transition is easier?

**Comment by Ericson2314** [comment]
Maybe we could agree to first land everything we need to generate config files based on flags and env vars, so the transition is easier?

I would still like to land this first, as it cleans up the code also and in such ways that I think make that config file generation a wee bit easier. We merge, but agree to revert this by the next release (I'll do the revert and fix conflicts) if the config file generation is not done in time.

**Comment by Ericson2314** [comment]
Ping. Anything we can do about this? Rebased for good measure, especiall now that that warning is fixed.

**Comment by ePirat** [comment]
IMO this is a good idea, as it will make mesons behavior match autotools and cmakes, and mean a lot less headache in complex cross-compile scenarios where meson, autotools and cmake projects coexist.

**Comment by jpakkane** [comment]
> Now `CC` always effects the host platform

This is not acceptable. The cross compiler must be defined in the cross file, only in the cross file and never, ever, under any circumstances anywhere else than the cross file. Doubly so for environment variables.

Taking the native compiler from CC_FOR_BUILD when cross compiling and CC when native compiling is probably ok, though, if I understood this change properly. I can't think of a case where it would break things.

**Comment by Ericson2314** [comment]
> Taking the native compiler from CC_FOR_BUILD when cross compiling and CC when native compiling is probably ok, though, if I understood this change properly. I can't think of a case where it would break things.

OK whew! :)

> This is not acceptable. The cross compiler must be defined in the cross file, only in the cross file and never, ever, under any circumstances anywhere else than the cross file. Doubly so for environment variables.

I can do this, but it will make the code a more complex, potentially a lot more complex. I don't understand why it's not OK to merely make the cross file *always* override the environment variables. C.f. native files, where we do just that (native file overwrites env vars) rather than banning env vars altogether.

**Comment by Ericson2314** [comment]
I am confused by the logs, should I just rebase?

**Comment by Ericson2314** [comment]
@jpakkane? Remember also that with the PR as is, there is no cross without a cross file--no env var can trigger cross on its own.

**Comment by jpakkane** [comment]
> I can do this, but it will make the code a more complex, potentially a lot more complex.

In what ways. The logic should go roughly the following FWICT. When you are looking for:

- a native compiler
- when cross compiling

then if `CC_FOR_BUILD` is set, use that for looking up the compiler. In all other cases use `CC`. Where would the complexity of this case come in this case?

> I don't understand why it's not OK to merely make the cross file always override the environment variables.

Because one of the main design principles of Meson has always been to not use environment variables whenever possible. They are a _terrible_ way of configuring anything and lead to weird bugs by, for example, running some commands from a shell that has different envvar settings from what you expected it to have.

**Comment by ePirat** [comment]
@jpakkane but meson is _already_ using env variables when cross compiling, except that it applies it to the native compiler and not like everyone else does to the cross compiler.

**Comment by ePirat** [comment]
Most of the headaches and lots of wasted time we had so far with projects using meson was exactly because of this inconsistency to what mostly all other build systems in existence do, everyone that ran into this was extremely surprised by this behavior of meson.

**Comment by ePirat** [comment]
Note that

> Taking the native compiler from CC_FOR_BUILD when cross compiling and CC when native compiling is probably ok, though, if I understood this change properly. I can't think of a case where it would break things.

does not make much sense when you apply all other env vars anyway to the native compiler, as it makes the existence of `CC_FOR_BUILD` absolutely pointless and things even more confusing and inconsistent…

**Comment by Ericson2314** [comment]
OK I've made this PR do what @jpakkane asked for, as it was far easier than I thought it was going to be. But I'll make another PR because I still do believe in that last bit, as gross as env vars are, for better support with my distro (Nixpkgs) and the reasons @ePirat says.

**Comment by Ericson2314** [comment]
@ePirat the comprise does help because at least we always interpet the env vars the same way autoconf does (no more build vs host mix up). The bummer is just that we surprisingly ignore env vars when autoconf wouldn't:

- Autoconf uses `CC` for cross compilation too
- Autoconf often uses `CC_FOR_BUILD` for native compilation too (in those cases, it is fine using a different compiler for `native: true`) for native and cross builds alike)

**Comment by ePirat** [comment]
The thing is, either you follow existing conventions with these env vars and use them always for the host compiler, which is how it should haven been from the beginning, or you don't.

But when you don't, reading other env vars (like `CC_FOR_BUILD`) that follow the convention that meson is ignoring, makes it super weird as it implies CC applies to host, and so would CFLAGS for example. Which is not the case for meson.

Sorry if I have a bit strong opinion here but I've probably wasted accumulated about half a week looking into all sorts of extremely weird to figure out build issues because of this behavior and explaining it to countless of people working on VLC or related projects…

**Comment by Ericson2314** [comment]
I'll open the follow-up PR for you to comment on :)

**Comment by ePirat** [comment]
Of course, as pointed out in other places already, another solution would be to just dont use the env vars at all when cross compiling, this way it would avoid breaking because it mistakenly applied things intended for the host compiler to the build compiler and such.

**Comment by Ericson2314** [comment]
@ePirat opened #6806

> Of course, as pointed out in other places already, another solution would be to just dont use the env vars at all when cross compiling

For me, this would be strictly worse. I, and the Nixpkgs community, do a lot of cross compiling of packages that upstream never tries to to cross compile. Every unnecessary difference between cross and native is a bitrotted build waiting to happen.

**Comment by Ericson2314** [comment]
Whew. All ready!

**Comment by Ericson2314** [comment]
Merged in 3a4388e51dee5e7e58f1e5ad4e60fb73b4aacf08

**Comment by hwti** [comment]
Is there a way to set the build platform flags on cross builds, which works with different meson versions ?
With 0.53.x, I only found `CFLAGS`, which would be used for the host machine on 0.54.0.
With 0.54.0, it will be `CFLAGS_FOR_BUILD`, or `c_args` in the native file.

**Comment by Ericson2314** [comment]
@hwti you can you the native file.

**Comment by hwti** [comment]
> @hwti you can you the native file, which you should prefer over env vars anyways.

On 0.53.2, the `[properties]` section from the native file doesn't work.
At least, `-Dbuild.c_args="-O2"` works on the command line.
On master, thanks to c2e65650, `c_args` and others can be set in the native file.

**Pull request closed without merging**